source-azure-blob-storage/main.go

```diff
@@ -0,0 +1,342 @@
+	MatchKeys          string         `json:"matchKeys"`
+	AscendingKeys      bool           `json:"ascendingKeys"`
+}
+
+type credentials struct {
+	AzureClientID       string `json:"azureClientID"`
+	AzureClientSecret   string `json:"azureClientSecret"`
+	AzureTenantID       string `json:"azureTenantID"`
+	AzureSubscriptionID string `json:"azureSubscriptionID"`
+	ConnectionString    string `json:"connectionString"`
+}
+
+func (c config) Validate() error {
+	if c.Credentials.ConnectionString != "" {
+		return nil
+	}
+	var requiredProperties = [][]string{
+		{"AzureTenantID", c.Credentials.AzureTenantID},
+		{"AzureClientID", c.Credentials.AzureClientID},
+		{"AzureClientSecret", c.Credentials.AzureClientSecret},
+		{"AzureSubscriptionID", c.Credentials.AzureSubscriptionID},
+		{"StorageAccountName", c.StorageAccountName},
+		{"ContainerName", c.ContainerName},
+	}
+
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+func (c config) DiscoverRoot() string {
+	return c.ContainerName
+}
+
+func (c config) RecommendedName() string {
+	return c.ContainerName
+}
+
+func (c config) FilesAreMonotonic() bool {
+	return c.AscendingKeys
+}
+
+func (c config) ParserConfig() *parser.Config {
+	return c.Parser
+}
+
+func (c config) PathRegex() string {
+	return c.MatchKeys
+}
+
+func newAzureBlobStore(cfg config) (*azureBlobStore, error) {
+	blobUrl := fmt.Sprintf("https://%s.blob.core.windows.net/", cfg.StorageAccountName)
+	var client *azblob.Client
+	var err error
+
+	if cfg.Credentials.ConnectionString != "" {
+		client, err = azblob.NewClientFromConnectionString(cfg.Credentials.ConnectionString, nil)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		credential, err := azidentity.NewClientSecretCredential(cfg.Credentials.AzureTenantID, cfg.Credentials.AzureClientID, cfg.Credentials.AzureClientSecret, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		client, err = azblob.NewClient(blobUrl, credential, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	store := &azureBlobStore{client: client, cfg: &cfg}
+
+	err = store.check(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	return store, err
+}
+
+type azureBlobStore struct {
+	client *azblob.Client
+	cfg    *config
+}
+
+// check verifies that we can list objects in the bucket and potentially read an object in
+// the bucket. This is done in a way that requires only s3:ListBucket and s3:GetObject permissions,
+// since these are the permissions required by the connector.
+func (az *azureBlobStore) check(ctx context.Context) error {
+	// All we care about is a successful listing rather than iterating on all objects
+
+	maxResults := int32(1)
+	listingOptions := azblob.ListBlobsFlatOptions{MaxResults: &maxResults}
+
+	pager := az.client.NewListBlobsFlatPager(az.cfg.ContainerName, &listingOptions)
+	if !pager.More() {
+		log.Printf("The container '%q' is empty", az.cfg.ContainerName)
```

Comment:
Use `logrus` for logging - we use it everywhere else. Also don't log this line at all because it's probably somewhat common and it would confuse users.